[Linux] Unique network namespaces for tests

XToripuru opened the pull request:
### Feature

Currently tests on Linux use the same network namespace names (app, tool, etc), this PR adds additional identifier (app-N, currently N = 0, etc.). Additionally there's a small code cleanup around tests.

This PR is 1st in a series of PRs to parallelize Linux test execution, unique namespace names are needed for the tests to not interfere which each other.

### Testing
CI passes on fork

XToripuru committed:
Indexed namespaces

XToripuru committed:
Refactor

XToripuru committed:
Refactor

XToripuru committed:
Refactor

XToripuru committed:
Use Application class instead of raw path for tests

XToripuru committed:
Fix

XToripuru committed:
Refactor

XToripuru committed:
Refactor

XToripuru committed:
Refactor

XToripuru committed:
Fix indent

XToripuru committed:
Import Path

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

XToripuru committed:
Add import

enkiusz committed:
implement suggested improvements:
- implement generic wrapping function for application objects
- overload RunSubprocess in IsolatedRunner

Signed-off-by: Maciej Grela <m.grela@samsung.com>

XToripuru committed:
Fix

restyled-commits committed:
Restyled by isort

enkiusz committed:
fix problems preventing the testsuite to start, --interface-id handling needs to be fixed still

Signed-off-by: Maciej Grela <m.grela@samsung.com>

XToripuru committed:
Remove unused

XToripuru committed:
Fix import

XToripuru committed:
Fix typo

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

XToripuru committed:
Fix adding args

arkq commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -264,7 +266,7 @@ def items_with_key(self):
                 # This path varies, however it is a fixed python tool so it may be ok
                 self.chip_tool_with_python_cmd,
-                os.path.basename(self.chip_tool_with_python_cmd[-1]),
+                os.path.basename(self.chip_tool_with_python_cmd.args[-1]),
```
I think that `self.chip_tool_with_python_path.name` should be here. Please double check.

XToripuru committed:
Fix paths

restyled-commits committed:
Restyled by autopep8

XToripuru committed:
Use tuple instead of list

XToripuru committed:
Fix

XToripuru committed:
Fix

XToripuru committed:
Fix

XToripuru committed:
Fix arg creation

restyled-commits committed:
Restyled by autopep8

XToripuru committed:
Fix

XToripuru committed:
Add executable bit to yamltest_with_matter_repl_tester.py

XToripuru committed:
Wrap chip-tool with python3

XToripuru committed:
Wrap chip-tool with python3 (fix)

XToripuru committed:
Wrap chip-tool with python3 (fix)

restyled-commits committed:
Restyled by autopep8

XToripuru committed:
Rerun CI

XToripuru committed:
Bugfix

XToripuru committed:
Fix OTA tests

restyled-commits committed:
Restyled by autopep8

XToripuru committed:
Fix typo

enkiusz committed:
Merge remote-tracking branch 'upstream/master' into linux-ns

XToripuru committed:
Bugfix

XToripuru committed:
Rename Application to Subprocess

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

XToripuru commented on scripts/tests/chiptest/runner.py:
```diff
@@ -118,23 +120,45 @@ def get(self):
+            # Try harder to avoid any stdout buffering in our tests
+            application = application.wrap_with(('stdbuf', '-o0', '-i0'))
+
+        logging.info('RunSubprocess starting application %s' % application)
```
Renamed to "Subprocess".

arkq commented on scripts/tests/chiptest/linux.py:
```diff
@@ -197,6 +199,9 @@ def terminate(self):
             self.run(command)
 
+    def wrap_in_namespace(self, subprocess: Subprocess):
+        return subprocess.wrap_with(("ip", "netns", "exec", "{}-{}".format(subprocess.kind, self.index)))
```
Running process in a namespace this way is very Linux specific, I think. I can imagine a more complicated method on some other platform (if there are any...). More generic way would be to provide a dedicated "executor" which will start the subprocess in the correct namespace (the `ip netns exec` should be an implementation detail within the "LinuxNamespaceSubprocess" class). Then, you could use that "executor" in the `runner.py::Runner.RunSubprocess()` instead of `subprocess.Popen`.

arkq commented on scripts/tests/chiptest/test_definition.py:
```diff
@@ -454,40 +456,51 @@ def Run(self, runner, apps_register, paths: ApplicationPaths, pics_file: str,
+
                 if dry_run:
-                    logging.info(" ".join(python_cmd))
+                    logging.info("{} {}".format(python_cmd.path, " ".join(python_cmd.args)))
```
What if there is some wrapper used? Why not `.to_cmd()` since you have such API?

enkiusz committed:
Don't alias module with function argument

Signed-off-by: Maciej Grela <m.grela@samsung.com>

XToripuru committed:
Use *args in Subprocess methods and fix logging Subprocess

XToripuru committed:
Change add_args to with_args and use *args instead of tuple

XToripuru committed:
Add Executors to Runner

restyled-commits committed:
Restyled by autopep8

XToripuru committed:
Fix typo with executors

XToripuru committed:
Remove unused import for Path

XToripuru committed:
Rename subprocess to subproc

XToripuru committed:
Rename Subprocess to SubprocessInfo

XToripuru committed:
Rename args

XToripuru committed:
Don't close pipes

XToripuru committed:
Fix path usage

XToripuru committed:
Move LinuxNamespacedExecutor to linux.py

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by ruff

enkiusz committed:
with_args and wrap_with cannot take a tuple argument anymore

Signed-off-by: Maciej Grela <m.grela@samsung.com>

XToripuru committed:
Add DarwinExecutor

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

restyled-commits committed:
Restyled by ruff

XToripuru committed:
Fix typo

XToripuru committed:
Make darwin.py executable

arkq commented on scripts/tests/chiptest/linux.py:
```diff
@@ -197,6 +199,19 @@ def terminate(self):
+    def __init__(self, ns):
+        self.ns = ns
+
+    def run(self, subproc: SubprocessInfo, stdin, stdout, stderr):
```
I would default `stdin`, `stdout`, etc. to the default value used by `subprocess.Popen`, but better yet, I would create API similar to `asyncio.create_subprocess_exec` where there is also `**kw` which is passed directly to `Popen()`.

enkiusz replied:
The `run` function needs to get a subprocess info object because the `kind` information needs to be passed to it implement namespaced runners. `stdin` et al. improved.

arkq commented on scripts/tests/chiptest/runner.py:
```diff
@@ -118,29 +119,55 @@ def get(self):
+    def __post_init__(self):
+        self.path = pathlib.Path(self.path)
+
+    def __repr__(self) -> str:
```
Why this is overwritten? The whole benefit of dataclass is that it provides `__init__` and `__repr__` automatically.

arkq commented on scripts/tests/run_darwin_framework_ota_test.py:
```diff
@@ -111,6 +111,13 @@ def cmd_run(context, darwin_framework_tool, ota_requestor_app, ota_data_file, ot
+
+    if ota_requestor_app is not None:
+        ota_requestor_app = SubprocessInfo(kind='app', path=ota_requestor_app
+                                           ).with_args('--otaDownloadPath', ota_destination_file)
```
There is no need to use `with_args()` here, just create `SubprocessInfo` with args.

enkiusz committed:
Without closing pipes the process hangs on exit

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Make DarwinExecutor accessible on sys.platform == darwin, improve executor selection logic

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz commented on scripts/tests/chiptest/linux.py:
```diff
@@ -197,6 +199,19 @@ def terminate(self):
         for command in self.COMMANDS_TERMINATE:
             self.run(command)
 
+    def wrap_in_namespace(self, subproc: SubprocessInfo):
```
I don't agree, moving this logic to the executor leaks namespace implementation internals (such as how the namespace name is created).

enkiusz committed:
Don't wrap chip-tool with python3, some post-review fixes

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Implement recommended logging usage pattern

Signed-off-by: Maciej Grela <m.grela@samsung.com>

enkiusz committed:
Merge branch 'master' into linux-ns

enkiusz committed:
Implement post-review changes

- use lists in TestDefinition.Run() to build commandline arguments
- provide default `std*` arguments for executors
- restrict `kind` values in SubprocessInfo to match network namespace identifiers
- improve dry_run logging code
- improve separation of concerns between IsolatedNetworkNamespace and LinuxNamespacedRunner

Signed-off-by: Maciej Grela <m.grela@samsung.com>

restyled-commits committed:
Restyled by isort

enkiusz committed:
make ruff happy

Signed-off-by: Maciej Grela <m.grela@samsung.com>

XToripuru committed:
Merge branch 'master' into linux-ns

XToripuru committed:
Fix typo _setup_app_link_up

enkiusz committed:
Merge branch 'master' of https://github.com/project-chip/connectedhomeip into linux-ns
